uqcsbot/starboard.py

```diff
@@ -0,0 +1,202 @@
+        self.starboard_channel = discord.utils.get(self.bot.get_all_channels(), name=self.CHANNEL_NAME)
+    
+    def _query_sb_message(self, recv: int) -> Optional[int]:
+        """ Get the starboard message ID corresponding to the recieved message """
+        db_session = self.bot.create_db_session()
+        id = db_session.query(models.Starboard).filter(models.Starboard.recv == recv).one_or_none()
+        db_session.close()
+
+        if id is not None:
+            return id.sent
+        return id
+
+    def _update_sb_message(self, recv: int, sent: int):
+        """ Sets the ID of the starboard message corresponding to the recieved message """
+        db_session = self.bot.create_db_session()
+        db_session.add(models.Starboard(recv=recv, sent=sent))
+        db_session.commit()
+        db_session.close()
+    
+    def _remove_sb_message(self, recv: int):
+        """ Deletes the DB entry for the starboard message for this recieved message """
+        db_session = self.bot.create_db_session()
+        entry = db_session.query(models.Starboard).filter(models.Starboard.recv == recv)
+        entry.delete(synchronize_session=False)
+        db_session.commit()
+        db_session.close()
+    
+    def _create_sb_embed(self, recv: discord.Message) -> discord.Embed:
+        embed = discord.Embed(color=recv.author.top_role.color, description=recv.content)
+        embed.set_author(name=recv.author.display_name, icon_url=recv.author.display_avatar.url)
+        embed.set_footer(text=recv.created_at.strftime('%b %d, %H:%M:%S'))
+
+        if len(recv.attachments) > 0:
+            embed.set_image(url = recv.attachments[0].url)
+            # only takes the first attachment to avoid sending large numbers of images to starboard.
+        
+        if recv.reference is not None and not isinstance(recv.reference, discord.DeletedReferencedMessage):            
+            replied = discord.Embed(color=recv.reference.resolved.author.top_role.color, description=recv.reference.resolved.content)
+            replied.set_author(name=f"Replying to {recv.reference.resolved.author.display_name}", icon_url=recv.reference.resolved.author.display_avatar.url)
+            replied.set_footer(text=recv.reference.resolved.created_at.strftime('%b %d, %H:%M:%S'))
+
+            return [replied, embed]
+        
+        return [embed]
+    
+    @commands.Cog.listener()
+    async def on_raw_reaction_add(self, payload: discord.RawReactionActionEvent):
+        if payload.emoji.id != self.starboard_emoji.id:
+            return
+        
+        guild = self.bot.get_guild(payload.guild_id)
+        if guild is None:
+            return
+        
+        channel = self.bot.get_channel(payload.channel_id)
+        if channel == self.starboard_channel or channel.category.name.startswith("admin"):
+            # TODO: "reaction count" for starboard could take into account (original + starboard)
+            return
```

Comment:
Comment: Bad future idea would be to have an admin category starboard

Reply:
I honestly prefer the idea of just making it the same starboard, to allow members a sneak preview of backstage shenanigans. It's easy enough to avoid using a :starhaj: react on "classified" info.

Alas, overruled by committee, but maybe after a bedding-in period we can demonstrate it, or enable it for admin-extra only.